### PR TITLE
Allow search information for a regions list

### PR DESCRIPTION
**Pull request by jesusbv**
As stated in the description, allow regions list
comma separated search

### CODE REVIEW

File: bin/pint [hunk 1]
```diff
@@ -150,15 +150,28 @@ if command_args['--region']:
...
-            framework,
-            image_state,
-            output_format,
-            region,
-            command_args['--filter'])
-        if '<images/>' in data or '"images": []' in data:
-            data = 'No information available. Please check your filter'
-        print(data)
+        no_info_regions = []
+        for reg in region.split(','):
```
Comment (smarlowucf) [inline bin/pint:154]: This seems to make region a required arg.

File: bin/pint [hunk 2]
```diff
@@ -150,15 +150,28 @@ if command_args['--region']:
...
+            else:
+                print(data)
+
+        if no_info_regions:
+            no_info_regions = (', ').join(no_info_regions)
+            data = "No information available for {} regions.".format(
+                no_info_regions
+            )
+            data += " Please check your filter"
+            print(data)
```
Comment (smarlowucf) [inline bin/pint:173]: If you choose multiple regions and one has no data while the others have data you end up with output in two different formats. An xml / json list followed by a string which makes the list not parsable.

**Review by smarlowucf — CHANGES_REQUESTED** [review]

File: bin/pint (hunk 1 above)
Comment (jesusbv) [reply bin/pint:154]: region is `all` by default
I do not understand what you mean

File: bin/pint (hunk 2 above)
Comment (jesusbv) [reply bin/pint:173]: OK, I think that the message telling the user to check the filter may need to be removed,
and PINT could return data as it is, i.e.
```
{
  "images": []
}
```
if empty and JSON format

Thoughts ?

File: bin/pint (hunk 1 above)
Comment (smarlowucf) [reply bin/pint:154]: Ack I did not see the default being set

File: bin/pint (hunk 2 above)
Comment (smarlowucf) [reply bin/pint:173]: IMHO if there's data we don't want an added warning message.

Comment (jesusbv) [reply bin/pint:173]: This change also would/should affect [regions](https://github.com/SUSE-Enceladus/public-cloud-info-client/blob/master/bin/pint#L197) 
```python
if '<regions/>' in regions or '"regions": []' in regions:
            regions = ('No region information available. Images have '
                       'the same identifier in all regions')
```

Comment (jesusbv) [reply bin/pint:173]: > IMHO if there's data we don't want an added warning message.

Agree but we should inform for a specific region there was no data, right ?
is the lack of response enough ?
let's say a user queries regions 'foo', 'bar' and 'foobar'
if 'bar' produces no data, we should return empty result, IMHO the lack of that info could lead to questions like
"did anything happen to the query ? to the response ?" an empty set says "query was OK but no result"

Comment (jesusbv) [reply bin/pint:195]: I edited the code so now, for example, it produces

```bash
<?xml version='1.0' encoding='UTF-8'?>
<servers><server ip="IP_A" ipv6="" name="domain" region="region_A" type="smt"/>
  <server ip="IP_B" ipv6="" name="domain" region="region_A" type="smt"/>
  <server ip="IP_C" ipv6="" name="domain" region="region_A" type="smt"/>
<server ip="IP_D" ipv6="" name="domain" region="region_B" type="smt"/>
  <server ip="IP_E" ipv6="" name="domain" region="region_B" type="smt"/>
  <server ip="IP_F" ipv6="" name="domain" region="region_B" type="smt"/>
</servers>
```
or a message to change the filters only if there is not a single region that returns no empty information 
```bash
> bin/pint microsoft servers --filter='name~foo' --json --region=region_a,region_b
No information available. Please check your filter
> 
```

File: bin/pint [hunk 3]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
-            output_format,
-            region,
-            command_args['--filter'])
-        if '<images/>' in data or '"images": []' in data:
-            data = 'No information available. Please check your filter'
-        print(data)
+        info = False
+        no_info = ''
+        data_json = {}
+        data_xml = None
```
Comment (rjschwei) [inline bin/pint:157]: I don't think any flags are needed. We take the users input and provide the information for that input. How would we even guess whether a region name is wrong or an image query is bogus?

File: bin/pint [hunk 4]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
+            data = ifsrequest.get_image_data(
+                framework,
+                image_state,
+                output_format,
+                reg.strip(),
+                command_args['--filter']
+            )
+            if '<images/>' in data or '"images": []' in data:
+                info = False
+                continue
```
Comment (rjschwei) [inline bin/pint:168]: OK, we do not add anything to what is eventually output when we get no information, no flag needed.

File: bin/pint [hunk 5]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
+
+            info = True
+            # there is info in some region/s
+            # so we show the empty response/s
+            # and not the message
+            if output_format in output_options:
```
Comment (rjschwei) [inline bin/pint:174]: Unnecessary nesting, `output_format` has a default as such:
```
if output_format == 'json':
.....
elif output_format == 'xml':
...
```
is sufficient

File: bin/pint [hunk 6]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
+                image_state,
+                output_format,
+                reg.strip(),
+                command_args['--filter']
+            )
+            if '<images/>' in data or '"images": []' in data:
+                info = False
+                continue
+
+            info = True
```
Comment (rjschwei) [inline bin/pint:170]: no flag

File: bin/pint [hunk 7]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
+            # there is info in some region/s
+            # so we show the empty response/s
+            # and not the message
+            if output_format in output_options:
+                if 'json' in output_format:
+                    data = json.loads(data)
+                    data_json['images'] = data_json.get('images', []) + data.get('images')
+                    data = data_json
```
Comment (rjschwei) [inline bin/pint:178]: data get over written in the next loop `data = ifsrequest.get_image_data( ...` on line 159 this assignment is superflous.

Also if `data_json` is properly initialized as `data_json{'images': []'}` the expression simplifies to `data_json['images'] += data.get('images')`

File: bin/pint [hunk 8]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
+                else:
+                    if data_xml is None:
+                        data_xml = etree.Element('images')
```
Comment (rjschwei) [inline bin/pint:181]: `data_xml` can be initialized to an etree outside of the loop without considerable cost. The size of an etree element is 64 bytes

```
from lxml import etree
import sys
data_xml = etree.Element('images')
sys.getsizeof(data_xml)
64
```

I don't think we have to worry about those 64 bytes when we have just pulled a lot more data over the wire that we are holding in memory.

File: bin/pint [hunk 9]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
+
+                    data = etree.fromstring(data.encode('utf-8'))
+                    data_xml.extend(data)
+                    data = etree.tostring(
+                        data_xml,
+                        xml_declaration=True,
+                        encoding='UTF-8',
+                        pretty_print=True
+                    ).decode()
```
Comment (rjschwei) [inline bin/pint:190]: This get overwritten in the next iteration of the loop on line 159 and is superfluous.

You need an assignment outside the loop to make a decision about what to print. Also when the output format is "plain" there is no accumulator, as such ```pint amazon --region us-east-1,eu-central-1 --active --filter 'name~byos'``` will fail to provide the expected information as the plain text data is not accumulated across the regions that are being processed.

File: bin/pint [hunk 10]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
+
+        if not info:
+            print("No information available. Please check your filter")
```
Comment (rjschwei) [inline bin/pint:193]: No message, we'd even print that message if the user provides no filter, i.e. ```pint amazon images --active --region a-new-region-with-no-images-yet``` would still produce a message telling the user to check the filter when they applied no filter.

File: bin/pint [hunk 11]
```diff
@@ -150,22 +151,94 @@ if command_args['--region']:
...
+        else:
+            print(data)
+
```
Comment (rjschwei) [inline bin/pint:240]: same issues apply to the server implementation

**Review by rjschwei — CHANGES_REQUESTED** [review]

**Comment by smarlowucf** [comment]
json output is no longer valid:

```
08:01:26 ▶ pint google images --active --json | jq
parse error: Invalid numeric literal at line 1, column 10
```

File: bin/pint (hunk 7 above)
Comment (jesusbv) [reply bin/pint:178]: if `data` or any other common variable for both cases (`json` and `xml`) when returning/printing the info after the loop,
there would have to be another check to print the output properly (json.dumps or etree.tostring, relatively), which I guess check again is no biggie

File: bin/pint (hunk 3 above)
Comment (jesusbv) [reply bin/pint]: The flags were initially to check if no info and print a message, but that is not the case anymore

**Comment by jesusbv** [comment]
If all OK, I would like to squash all the commits into one before merging

**Review by smarlowucf — APPROVED** [review]
IMHO this seems like a lot of complexity for not much value. In the end the user will still have to filter or group on region anyway. But as long as there's no more regressions :shrug: .

This brings with it other idiosyncrasies though. For example Google can end up with data duplication:

```
pint google images --active --region us-west1,us-east1 --filter="name~sapcal" --json | jq
{
  "images": [
    {
      "changeinfo": "https://publiccloudimagechangeinfo.suse.com/google/sles-15-sp4-sapcal-v20220722-x86-64/",
      "deletedon": "",
      "deprecatedon": "",
      "name": "sles-15-sp4-sapcal-v20220722-x86-64",
      "project": "suse-sap-cal-cloud",
      "publishedon": "20220722",
      "replacementname": "",
      "state": "active"
    },
    {
      "changeinfo": "https://publiccloudimagechangeinfo.suse.com/google/sles-15-sp4-sapcal-v20220722-x86-64/",
      "deletedon": "",
      "deprecatedon": "",
      "name": "sles-15-sp4-sapcal-v20220722-x86-64",
      "project": "suse-sap-cal-cloud",
      "publishedon": "20220722",
      "replacementname": "",
      "state": "active"
    }
  ]
}
```

**Comment by jesusbv** [comment]
> IMHO this seems like a lot of complexity for not much value. In the end the user will still have to filter or group on region anyway. But as long as there's no more regressions shrug .
> 
> This brings with it other idiosyncrasies though. For example Google can end up with data duplication:
> 
> ```
> pint google images --active --region us-west1,us-east1 --filter="name~sapcal" --json | jq
> {
>   "images": [
>     {
>       "changeinfo": "https://publiccloudimagechangeinfo.suse.com/google/sles-15-sp4-sapcal-v20220722-x86-64/",
>       "deletedon": "",
>       "deprecatedon": "",
>       "name": "sles-15-sp4-sapcal-v20220722-x86-64",
>       "project": "suse-sap-cal-cloud",
>       "publishedon": "20220722",
>       "replacementname": "",
>       "state": "active"
>     },
>     {
>       "changeinfo": "https://publiccloudimagechangeinfo.suse.com/google/sles-15-sp4-sapcal-v20220722-x86-64/",
>       "deletedon": "",
>       "deprecatedon": "",
>       "name": "sles-15-sp4-sapcal-v20220722-x86-64",
>       "project": "suse-sap-cal-cloud",
>       "publishedon": "20220722",
>       "replacementname": "",
>       "state": "active"
>     }
>   ]
> }
> ```

Good point, to keep it as it was, this is:
```bash
# production PINT
> pint google images --active  --filter="name~sapcal" --xml
<?xml version='1.0' encoding='UTF-8'?>
<images>
  <image changeinfo="https://publiccloudimagechangeinfo.suse.com/google/sles-15-sp4-sapcal-v20220722-x86-64/" deletedon="" deprecatedon="" name="sles-15-sp4-sapcal-v20220722-x86-64" project="suse-sap-cal-cloud" publishedon="20220722" replacementname="" state="active"/>
</images>

> 
> pint google images --active  --filter="name~sapcal" --json | jq
{
  "images": [
    {
      "changeinfo": "https://publiccloudimagechangeinfo.suse.com/google/sles-15-sp4-sapcal-v20220722-x86-64/",
      "deletedon": "",
      "deprecatedon": "",
      "name": "sles-15-sp4-sapcal-v20220722-x86-64",
      "project": "suse-sap-cal-cloud",
      "publishedon": "20220722",
      "replacementname": "",
      "state": "active"
    }
  ]
}
> 
```
Now, the output removes duplicates producing
```bash
# branch PINT
> bin/pint google images --active --region us-west1,us-east1,europe-west1 --filter="name~sapcal" --xml
<?xml version='1.0' encoding='UTF-8'?>
<images><image changeinfo="https://publiccloudimagechangeinfo.suse.com/google/sles-15-sp4-sapcal-v20220722-x86-64/" deletedon="" deprecatedon="" name="sles-15-sp4-sapcal-v20220722-x86-64" project="suse-sap-cal-cloud" publishedon="20220722" replacementname="" state="active"/>
</images>

> 
> bin/pint google images --active --region us-west1,us-east1,europe-west1 --filter="name~sapcal" --json | jq
{
  "images": [
    {
      "changeinfo": "https://publiccloudimagechangeinfo.suse.com/google/sles-15-sp4-sapcal-v20220722-x86-64/",
      "deletedon": "",
      "deprecatedon": "",
      "name": "sles-15-sp4-sapcal-v20220722-x86-64",
      "project": "suse-sap-cal-cloud",
      "publishedon": "20220722",
      "replacementname": "",
      "state": "active"
    }
  ]
}
> 
```

File: bin/pint [hunk 12]
```diff
@@ -150,22 +151,93 @@ if command_args['--region']:
...
+            if '<images/>' in data or '"images": []' in data:
+                continue
+
+            if 'json' in output_format:
+                data = json.loads(data)
+                if data.get('images')[0] not in data_json['images']:
+                    # add new elements only
+                    data_json['images'] += data.get('images')
+            elif output_format in ['xml', 'plain']:
+                data_content = data.partition("<images>")[-1][:-10].strip()
```
Comment (rjschwei) [inline bin/pint:175]: Magic slice numbers deserve a comment, what is `-10` cutting off?

**Review by rjschwei — CHANGES_REQUESTED** [review]

File: bin/pint (hunk 12 above)
Comment (jesusbv) [reply bin/pint]: Done

**Comment by rjschwei** [comment]
Going to close this. It has been sitting around for a long time and we've not had people beating down our door for this feature. Also we'd have to rename the `region` argument to `regions`, so at best this is only partially done. man pages would also need updates etc.

**Pull request closed without merging**